ynr/apps/candidates/views/people.py

```diff
@@ -194,48 +194,140 @@ def post(self, request, *args, **kwargs):
         )
 
 
-class MergePeopleView(GroupRequiredMixin, View):
+class MergePeopleMixin:
+    def do_merge(self, primary_person, secondary_person):
+        merger = PersonMerger(
+            primary_person, secondary_person, request=self.request
+        )
 
-    http_method_names = ["post"]
+        return merger.merge(delete=True)
+
+
+class MergePeopleView(GroupRequiredMixin, TemplateView, MergePeopleMixin):
+
+    http_method_names = ["get", "post"]
     required_group_name = TRUSTED_TO_MERGE_GROUP_NAME
+    template_name = "candidates/generic-merge-error.html"
 
-    def post(self, request, *args, **kwargs):
-        # Check that the person IDs are well-formed:
-        primary_person_id = self.kwargs["person_id"]
-        secondary_person_id = self.request.POST["other"]
-        if not re.search("^\d+$", secondary_person_id):
+    def validate(self, context):
+        if not re.search(r"^\d+$", context["other_person_id"]):
             message = "Malformed person ID '{0}'"
-            raise ValueError(message.format(secondary_person_id))
-        if primary_person_id == secondary_person_id:
+            raise ValueError(message.format(context["other_person_id"]))
+        if context["person"].pk == int(context["other_person_id"]):
             message = "You can't merge a person ({0}) with themself ({1})"
             raise ValueError(
-                message.format(primary_person_id, secondary_person_id)
+                message.format(context["person"].pk, context["other_person_id"])
             )
 
+    def get_context_data(self, **kwargs):
+        context = super().get_context_data(**kwargs)
+        context["person"] = get_object_or_404(
+            Person, id=self.kwargs["person_id"]
+        )
+        context["other_person_id"] = self.request.POST.get("other", "")
+        return context
+
+    def post(self, request, *args, **kwargs):
+        context = self.get_context_data()
+        # Check that the person IDs are well-formed:
+        try:
+            self.validate(context)
+        except ValueError as e:
```

Comment:
The problem with catching `ValueError` here is that if something else throws another `ValueError` (other than "you can't merge a person with themself" or "Malformed person ID"), we'll show the user the error and not log it to Sentry. That means if we throw a `ValueError` because our code is bad (e.g: `ValueError: need more than 3 values to unpack`, or whatever), we'll never find out about it unless a user tells us.

I think we need a special exception class for stuff where we want to tell the user "you're doing it wrong" and differentiate that from runtime errors (which we want notifications about).

Reply:
👍 Switched to `InvalidMergeError`